src/contracts.rs

```diff
@@ -109,44 +131,86 @@ impl TonContracts {
     /// Get address for contract deploying
     pub fn get_deploy_address(
         &self,
-        code: &[u8],
-        keys: &Ed25519KeyPair,
+        abi: &str,
+        image: &[u8],
+        init_params: Option<RunParameters>,
+        public_key: &Ed25519Public,
+        workchain_id: i32,
     ) -> TonResult<TonAddress> {
-        let result: TonAddress = Interop::json_request(
+        self.get_deploy_data(Some(abi), Some(image), init_params, public_key, Some(workchain_id))?
+            .address
+            .ok_or(TonErrorKind::InternalError("No address in result".to_owned()).into())
+    }
+
+    /// Get contract deploy data: image (state init), storage data and deploying address
+    pub fn get_deploy_data(
+        &self,
+        abi: Option<&str>,
+        image: Option<&[u8]>,
+        init_params: Option<RunParameters>,
+        public_key: &Ed25519Public,
+        workchain_id: Option<i32>,
+    ) -> TonResult<ResultOfGetDeployData> {
+        let abi = abi.map(|val| 
+                serde_json::from_str(val).map_err(|_| TonErrorKind::InvalidArg(val.to_owned())))
+            .transpose()?;
+
+        let core_result: ResultOfGetDeployDataCore = Interop::json_request(
             self.context,
-            "contracts.deploy.address",
-            ParamsOfGetDeployAddress {
-                abi: Value::Null,
-                imageBase64: base64::encode(code),
-                keyPair: keys.clone(),
+            "contracts.deploy.data",
+            ParamsOfGetDeployData {
+                abi,
+                imageBase64: image.map(|val| base64::encode(val)),
+                initParams: Self::option_params_to_value(init_params)?,
+                publicKeyHex: public_key.clone(),
+                workchainId: workchain_id,
             })?;
-        Ok(result)
+
+        Ok(ResultOfGetDeployData {
+            address: core_result.address.map(|val| TonAddress::from_str(&val)).transpose()?,
+            image: core_result.imageBase64.map(|val| base64::decode(&val).into()).transpose()?,
+            data: base64::decode(&core_result.dataBase64)?
+        })
+    }
+
+    fn params_to_value(params: RunParameters) -> TonResult<Value> {
+        let str_params = match &params {
+            RunParameters::Json(string) => string
+        };
+       serde_json::from_str(str_params)
+            .map_err(|_| TonErrorKind::InvalidArg(str_params.to_owned()).into())
+    }
+
+    fn option_params_to_value(params: Option<RunParameters>) -> TonResult<Option<Value>> {
+        match params {
+            Some(params) => Ok(Some(Self::params_to_value(params)?)),
+            None => Ok(None)
+        }
     }
 
     /// Deploy contract to TON blockchain
     pub fn deploy(
         &self,
         abi: &str,
         code: &[u8],
+        constructor_header: Option<RunParameters>,
```

Comment:
Why not just a `header`? or `abi_header`?

Reply:
It contains header params for constructor call. Like `constructor_params` contains parameters for constructor itself